ilastik/applets/counting/countingGuiBoxesInterface.py

```diff
@@ -970,41 +973,39 @@ def _getNextBoxColor(self):
         return color
 
     def saveBoxesToCSV(self, filename):
-        import os, csv
-
-        b, ext = os.path.splitext(str(filename))
-        assert ext == ".txt", "wrong filename or extension %s" % str(filename)
         try:
-            with open(filename, "wb") as fh:
+            with open(filename, "w", newline="") as csvfile:
+                writer = csv.writer(csvfile, quoting=csv.QUOTE_MINIMAL)
+                writer.writerow(["ID", "StartX", "StartY", "StopX", "StopY", "Count", "Average density", "Std density"])
+
+                for row_id, box in enumerate(self._currentBoxesList):
+                    start = [str(p) for p in box.getRectItem().topLeftDataPos()]
+                    stop = [str(p + 1) for p in box.getRectItem().bottomRightDataPos()]
+                    region = box.getSubRegion()
+                    count = format(np.sum(region), "0.8g")
+                    avg_density = format(np.mean(region), "0.8g")
+                    std_density = format(np.std(region), "0.8g")
+                    writer.writerow([str(row_id), *start, *stop, count, avg_density, std_density])
 
-                header = ["ID", "StartX", "StartY", "StopX", "StopY", "Count", "Average density", "Std density"]
+        except IOError as e:
+            logger.error(e)
+            raise
 
-                fh.write(" , ".join(header) + "\n")
+    def loadBoxesFromCSV(self, filename):
+        colnames = "StartX", "StartY", "StopX", "StopY"
 
-                for k, box in enumerate(self._currentBoxesList):
-                    start = box.getStart()
-                    stop = box.getStop()
-                    region = box.getSubRegion()
-                    count = np.sum(region)
-                    averagedens = np.mean(region)
-                    stddensity = np.std(region)
-
-                    line = [
-                        "%5.5d" % k,
-                        "%5.5d" % start[1],
-                        "%5.5d" % start[2],
-                        "%5.5d" % stop[1],
-                        "%5.5d" % stop[2],
-                        "%5.2f" % count,
-                        "%5.2f" % averagedens,
-                        "%5.2f" % stddensity,
-                    ]
-                    logger.debug("line " + ",".join(line))
-                    fh.write(",".join(line) + "\n")
+        try:
+            with open(filename, newline="") as csvfile:
+                for row in csv.DictReader(csvfile):
+                    cols = [int(row[name]) for name in colnames]
+                    self.addNewBox(roi2rect(cols[:2], cols[2:]))
```

Comment:
I guess a bit nitpicky, but it's possible that we throw an exception after having added a few boxes. It might be safer to first collect and validate all boxes, then add all of them at once (or rollback, if you're feeling adventurous) so that we can't reach weird states of semi-imported files.

Reply:
Now we collect all boxes from CSV first, and only then actually add them into view/controller.

For export, we just remove the file if something went wrong.